### PR TITLE
Feature/multi tag search

### PR DESCRIPTION
**Pull request by boubaker**

### CODE REVIEW

File: core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java [hunk 1]
```diff
@@ -302,6 +305,60 @@ public void addSiteTag(String siteName,
...
+      firstSearch = false;
+    }
+    Iterator<Node> nodesIterator = ret.iterator();
+    while (nodesIterator.hasNext()) {
+      Node node = nodesIterator.next();
+      if (!nodesPaths.contains(node.getPath())) {
+        nodesIterator.remove();
+      }
+    }
+    return ret;
```
Comment (thomasdelhomenie) [inline core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java]: Just a thought, the current algo loops over all tags, then all nodes of the tags to gather the nodes paths and then filter them. Maybe it would be more efficient to fetch the nodes of the first tag (as it is currently done), then loop over these nodes to check if they have the others tags. It would avoid to fetch a lot of nodes if tags are applied on a lot of nodes. WDYT ?

File: core/services/src/test/java/org/exoplatform/services/ecm/dms/folksonomy/TestNewFolksonomyService.java [hunk 2]
```diff
@@ -343,6 +346,71 @@ public void testAddSiteTag() throws Exception {
...
   /**
+   * Test Method: getAllDocumentsIteratorByTagsAndPath()
+   * Input: Nodes 'test', 'test2'
+   * Test action: add 2 tags 'sport' and 'weather' for node 'test' and 'test2' of current user
+   *               get all documents by tag 'sport'
+   * Expected Result:
+   *              'test' and 'test2'
+   */
+  
+  public void testGetAllDocumentsIteratorByTagsAndPath() throws Exception {
```
Comment (thomasdelhomenie) [inline core/services/src/test/java/org/exoplatform/services/ecm/dms/folksonomy/TestNewFolksonomyService.java]: It seems this test only uses 1 tag when calling getAllDocumentsIteratorByTagsAndPath (weatherTagNode is never used). We should add a call with the 2 tags IMO.
Also we should add a case where a document has only 1 tag, then fetch with this tag (it should be returned), then fetch with the 2 tags (it should not be returned). WDYT ?

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/control/UIActionBar.java [hunk 3]
```diff
@@ -23,6 +23,7 @@
 import org.exoplatform.ecm.webui.component.explorer.popup.actions.UIDocumentForm;
 import org.exoplatform.ecm.webui.component.explorer.popup.actions.UIDocumentFormController;
 import org.exoplatform.ecm.webui.component.explorer.search.*;
+import org.exoplatform.ecm.webui.component.explorer.sidebar.UITagExplorer;
```
Comment (thomasdelhomenie) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/control/UIActionBar.java]: Useless import

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/sidebar/UITagExplorer.java [hunk 4]
```diff
@@ -89,7 +93,8 @@ public UITagExplorer() throws Exception {
   }
 
   public List<Node> getPublicTagLink() throws Exception {
-    return NodeLocation.getNodeListByLocationList(publicTagPageIterator_.getCurrentPageData());
+    List nodeListByLocationList = NodeLocation.getNodeListByLocationList(publicTagPageIterator_.getCurrentPageData());
+    return nodeListByLocationList;
```
Comment (thomasdelhomenie) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/sidebar/UITagExplorer.java]: Just curious, why this change ?

File: core/services/src/main/java/org/exoplatform/services/cms/folksonomy/NewFolksonomyService.java [hunk 5]
```diff
@@ -167,6 +168,23 @@ public void addSiteTag(String siteName,
...
+   * @param selectedPath Parent path to filter nodes
+   * @param tagPaths_ list of tags JCR paths
+   * @param workspace the workspace of resulted nodes
+   * @param sessionProvider use session provider to query JCR
+   * 
+   * @return the filtered {@link List} of {@link Node}
+   * 
+   * @throws Exception
+   */
+  List<Node> getAllDocumentsIteratorByTagsAndPath(String selectedPath,
```
[the remark this answers is not in the document]
Comment (boubaker) [reply core/services/src/main/java/org/exoplatform/services/cms/folksonomy/NewFolksonomyService.java]: renamed

File: core/services/src/main/java/org/exoplatform/services/cms/folksonomy/NewFolksonomyService.java [hunk 6]
```diff
@@ -167,6 +168,23 @@ public void addSiteTag(String siteName,
...
+                                           Set<String> tagPaths_,
```
[the remark this answers is not in the document]
Comment (boubaker) [reply core/services/src/main/java/org/exoplatform/services/cms/folksonomy/NewFolksonomyService.java]: renamed

File: core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java [hunk 7]
```diff
@@ -302,6 +305,60 @@ public void addSiteTag(String siteName,
...
    */
+  @Override
+  public List<Node> getAllDocumentsIteratorByTagsAndPath(String selectedPath,
+                                                  Set<String> tagPaths,
+                                                  String workspace,
+                                                  SessionProvider sessionProvider) throws Exception {
+    List<Node> ret = new ArrayList<Node>();
+    Set<String> nodesPaths = new HashSet<>();
+    boolean firstSearch = true;
+    for (String tagPath : tagPaths) {
```
[the remark this answers is not in the document]
Comment (boubaker) [reply core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java]: Check added

File: core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java [hunk 8]
```diff
@@ -302,6 +305,60 @@ public void addSiteTag(String siteName,
...
+          try {
+            targetNode = linkManager.getTarget(node);
+          } catch (Exception e) {
+            if (LOG.isWarnEnabled()) {
+              LOG.warn(e.getMessage());
+            }
+          }
+          if (targetNode != null && !((Node) targetNode.getAncestor(1)).isNodeType(EXO_TRASH_FOLDER)
+              && targetNode.getSession().getWorkspace().getName().equals(workspace)
+              && (StringUtils.isBlank(selectedPath) || targetNode.getPath().startsWith(selectedPath))) {
```
[the remark this answers is not in the document]
Comment (boubaker) [reply core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java]: condition checked

File: core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java (hunk 1 above)
Comment (boubaker) [reply core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java]: > the current algo loops over all tags

It loops only on selected tags

> It would avoid to fetch a lot of nodes if tags are applied on a lot of nodes

I don't think so. In fact, to get the selected tags of each node, we will have to query all collaboration workspace to get all tagged documents and then filter on them.
Or if you want to test if a symlink of a document exists in tag node, you will iterate anyway on all documents of each tag for each selected node.
This is caused by the fact that in JCR, we don't have the opposite reference when using symlink in target node.

File: core/services/src/test/java/org/exoplatform/services/ecm/dms/folksonomy/TestNewFolksonomyService.java (hunk 2 above)
Comment (boubaker) [reply core/services/src/test/java/org/exoplatform/services/ecm/dms/folksonomy/TestNewFolksonomyService.java]: additional test added

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/control/UIActionBar.java (hunk 3 above)
Comment (boubaker) [reply core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/control/UIActionBar.java]: class import deleted

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/sidebar/UITagExplorer.java (hunk 4 above)
Comment (boubaker) [reply core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/sidebar/UITagExplorer.java]: changed line reverted

File: core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java (hunk 1 above)
Comment (thomasdelhomenie) [reply core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java]: ok :+1: . This is probably something to improve in the future to have good performances, but it requires some changes in the  design and must be done in a dedicated task.
